libp2p/crypto/ed25519/ed25519.nim

```diff
@@ -1638,34 +1643,41 @@ proc checkScalar*(scalar: openarray[byte]): uint32 =
     c = -1
   result = NEQ(z, 0'u32) and LT0(c)
 
-proc random*(t: typedesc[EdPrivateKey]): EdPrivateKey =
+proc random*(t: typedesc[EdPrivateKey]): Result[EdPrivateKey, EdError] =
   ## Generate new random ED25519 private key using OS specific CSPRNG.
   var
     point: GeP3
     pk: array[EdPublicKeySize, byte]
-  if randomBytes(result.data.toOpenArray(0, 31)) != 32:
-    raise newException(EdRngError, "Could not generate random data")
-  var hh = sha512.digest(result.data.toOpenArray(0, 31))
-  hh.data[0] = hh.data[0] and 0xF8'u8
-  hh.data[31] = hh.data[31] and 0x3F'u8
-  hh.data[31] = hh.data[31] or 0x40'u8
-  geScalarMultBase(point, hh.data)
-  geP3ToBytes(pk, point)
-  copyMem(addr result.data[32], addr pk[0], 32)
-
-proc random*(t: typedesc[EdKeyPair]): EdKeyPair =
+    res: EdPrivateKey
+  if randomBytes(res.data.toOpenArray(0, 31)) != 32:
+    err(EdRngError)
+  else:
+    var hh = sha512.digest(res.data.toOpenArray(0, 31))
+    hh.data[0] = hh.data[0] and 0xF8'u8
+    hh.data[31] = hh.data[31] and 0x3F'u8
+    hh.data[31] = hh.data[31] or 0x40'u8
+    geScalarMultBase(point, hh.data)
+    geP3ToBytes(pk, point)
+    copyMem(addr res.data[32], addr pk[0], 32)
+    ok(res)
+
+proc random*(t: typedesc[EdKeyPair]): Result[EdKeyPair, EdError] =
   ## Generate new random ED25519 private and public keypair using OS specific
   ## CSPRNG.
-  var point: GeP3
-  if randomBytes(result.seckey.data.toOpenArray(0, 31)) != 32:
-    raise newException(EdRngError, "Could not generate random data")
-  var hh = sha512.digest(result.seckey.data.toOpenArray(0, 31))
-  hh.data[0] = hh.data[0] and 0xF8'u8
-  hh.data[31] = hh.data[31] and 0x3F'u8
-  hh.data[31] = hh.data[31] or 0x40'u8
-  geScalarMultBase(point, hh.data)
-  geP3ToBytes(result.pubkey.data, point)
-  copyMem(addr result.seckey.data[32], addr result.pubkey.data[0], 32)
+  var
+    point: GeP3
+    res: EdKeyPair
+  if randomBytes(res.seckey.data.toOpenArray(0, 31)) != 32:
+    err(EdRngError)
+  else:
+    var hh = sha512.digest(res.seckey.data.toOpenArray(0, 31))
+    hh.data[0] = hh.data[0] and 0xF8'u8
+    hh.data[31] = hh.data[31] and 0x3F'u8
+    hh.data[31] = hh.data[31] or 0x40'u8
+    geScalarMultBase(point, hh.data)
+    geP3ToBytes(res.pubkey.data, point)
+    copyMem(addr res.seckey.data[32], addr res.pubkey.data[0], 32)
```

Comment:
not for this branch probably, but `copyMem` like these can usually be done as `res.seckey.data[32..<64] = res.pubkey.data`

Reply:
but your version would `Defect` as well and probably be slower too I think (not memcpy) in this case

Reply:
not really - the = operator can be convinced to use copyMem - as long as the lengths check out, it will not defect (so yeah, you still need to do a length check)